src/pam_google_authenticator.c

```diff
@@ -385,6 +392,38 @@ static int drop_privileges(pam_handle_t *pamh, const char *username, int uid,
   gid_t gid = pw->pw_gid;
   free(buf);
 
+  // to access the secret file, we may have to be member of the same groups as the 
+  // user that tries to authenticate.
+  // get number of groups for the user
+  GroupList user_groups;
+  user_groups.n_groups = 0;
+  int err = getgrouplist(username, gid, NULL, &user_groups.n_groups);
+
+  // now get a full list of all groups
+  user_groups.gids = malloc(sizeof(gid_t) * user_groups.n_groups);
+  err = getgrouplist(username, gid, user_groups.gids, &user_groups.n_groups);
+  if (err == -1) {
+    log_message(LOG_ERR, pamh, "Failed to get group memberships of user \"%s\"", username);
+  }
+
+  // store the old list of groups of this process
+  old_groups->n_groups = getgroups(0, NULL);
+  old_groups->gids = malloc(sizeof(gid_t)*old_groups->n_groups);
+  err = getgroups(old_groups->n_groups, old_groups->gids);
+  if (err == -1) {
+    log_message(LOG_ERR, pamh, "Failed to get supplementary groups for the process");
+  }
+
+  // change the list of groups to the groups the user is member of
+  err = setgroups(user_groups.n_groups, user_groups.gids);
+  if (err == -1) {
+    log_message(LOG_ERR, pamh, "Failed to set %d supplementary group IDs for the process", user_groups.n_groups);
+  }
+  // cleanup user group list
+  free(user_groups.gids);
+  user_groups.n_groups = -1;
```

Comment:
why not 0?

Well, if this is in its own function then this auto variable will go out of scope anyway.